### PR TITLE
LC-470: create FileConnector

### PR DESCRIPTION
**Pull request by Jozurf**
The FileConnector can connect to the local filesystem, google cloud storage, s3 storage and azure storage depending on the path given to the pathToStorage configuration.

Initially was planning to use each of the cloud providers Java NIO package to connect to cloud storages. This would allow us to get their respective FileSystems easily in one codebase and traverse all of them. However, after some testing, the individual sdk's trumps their respective NIO package (3-4 times better performance). After some discussion and exploration, decided that it was possible to use their own sdk by creating a custom interface and abstract class to extract repeated code across the sdks.

### CODE REVIEW

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+ *    "accountKey" : azure account key
+ *  Amazon:
+ *    "accessKeyId" : s3 key id
+ *    "secretAccessKey" : secret access key
+ *    "region" : s3 storage region
+ *
+ * Optional:
+ *  "maxNumOfPages" : number of reference to the files loaded into memory in a single fetch request, defaults to 100
+ */
+
```
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java:44]: extra space

Comment (Jozurf) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java]: removed in [1082fb3](https://github.com/kmwtechnology/lucille/pull/205/commits/1082fb30901ad893f413e5243c62ff2ec98b48d5)

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/cloudstorageclients/CloudStorageClient.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package com.kmwllc.lucille.connector.cloudstorageclients;
+
+import com.kmwllc.lucille.core.Publisher;
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+public interface CloudStorageClient {
```
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/cloudstorageclients/CloudStorageClient.java:9]: would be good to have some doc on what this interface is and the individual methods it includes

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/cloudstorageclients/GoogleStorageClient.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      doc.setField(FileConnector.SIZE, blob.getSize());
+      doc.setField(FileConnector.CONTENT, blob.getContent());
+    } catch (Exception e) {
+      throw new IOException("Error occurred getting/setting file attributes to document: " + blob.getName(), e);
+    }
+    return doc;
+  }
+
+  // Only for testing
+  public void setStorageForTesting(Storage storage) {
```
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/cloudstorageclients/GoogleStorageClient.java:95]: try making this package-private (leave off the "public" access modifier) -- should still be accessible in the test assuming the test is in the same package

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/cloudstorageclients/S3StorageClient.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+  private boolean isNotValid(S3Object obj) {
+    String key = obj.key();
+    if (key.endsWith("/")) return true;
+
+    return shouldSkipBasedOnRegex(key);
+  }
+
+  // Only for testing
+  public void setS3ClientForTesting(S3Client s3) {
```
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/cloudstorageclients/S3StorageClient.java:91]: try making package-private

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/cloudstorageclients/GoogleStorageClient.java (hunk 3 above)
Comment (Jozurf) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/cloudstorageclients/GoogleStorageClient.java]: resolved in [41ef875](https://github.com/kmwtechnology/lucille/pull/205/commits/41ef8752a6d87d285d70ae2a101ea75109612c46)

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/cloudstorageclients/CloudStorageClient.java (hunk 2 above)
Comment (Jozurf) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/cloudstorageclients/CloudStorageClient.java:20]: updated in [41ef875](https://github.com/kmwtechnology/lucille/pull/205/commits/41ef8752a6d87d285d70ae2a101ea75109612c46)

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+      validateCloudOptions(storageURI, cloudOptions);
+      cloudOptions.put(GET_FILE_CONTENT, getFileContent);
+      cloudStorageClient = CloudStorageClient.getClient(storageURI, publisher, getDocIdPrefix(), excludes, includes, cloudOptions);
+      cloudStorageClient.init();
+      cloudStorageClient.publishFiles();
+      try {
+        cloudStorageClient.shutdown();
+      } catch (Exception e) {
+        throw new ConnectorException("Failed to shutdown client.", e);
+      }
```
Comment (rseitz) [inline lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java:104]: could return here and remove the else block on l105 to reduce nesting

Comment (Jozurf) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java:109]: updated in [b77e41f](https://github.com/kmwtechnology/lucille/pull/205/commits/b77e41f1e197220437a1fe6654b8194e6be92412)

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    }
+  }
+
+  @Override
+  public void execute(Publisher publisher) throws ConnectorException {
+    if (storageURI.getScheme() != null) {
+      validateCloudOptions(storageURI, cloudOptions);
+      cloudOptions.put(GET_FILE_CONTENT, getFileContent);
+      cloudStorageClient = CloudStorageClient.getClient(storageURI, publisher, getDocIdPrefix(), excludes, includes, cloudOptions);
+      cloudStorageClient.init();
```
[the remark this answers is not in the document]
Comment (Jozurf) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java]: resolved in [b77e41f](https://github.com/kmwtechnology/lucille/pull/205/commits/b77e41f1e197220437a1fe6654b8194e6be92412)

**Review by rseitz — APPROVED** [review]

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+ *    "accountName" : azure account name
+ *    "accountKey" : azure account key
+ *  Amazon:
+ *    "accessKeyId" : s3 key id
+ *    "secretAccessKey" : secret access key
+ *    "region" : s3 storage region
+ *
+ * Optional:
+ *  "maxNumOfPages" : number of reference to the files loaded into memory in a single fetch request, defaults to 100
+ */
```
[the remark this answers is not in the document]
Comment (Jozurf) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java:49]: added config properties to javadocs in [c661fd0](https://github.com/kmwtechnology/lucille/pull/205/commits/c661fd05b3643eb47f8a638c42e5b0b2613bd7e9)

**Review by spencersolomon6 — APPROVED** [review]

File: lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (spencer-solomon-tfs) [reply lucille-core/src/main/java/com/kmwllc/lucille/connector/FileConnector.java:54]: Ticket has been created for this: LC-569

**Pull request merged**